### PR TITLE
Add shape checks to tell() methods

### PR DESCRIPTION
**Pull request by itsdawei**
## Description

Add shape checks to `tell()` and `tell_dqd()`.

## TODO

- [x] Add shape checks
- [x] Add tests

## Status

- [x] I have read the guidelines in [CONTRIBUTING.md](https://github.com/icaros-usc/pyribs/blob/master/CONTRIBUTING.md)
- [x] I have formatted my code using `yapf`
- [x] I have tested my code by running `pytest`
- [x] I have linted my code with `pylint`
- [x] I have added a one-line description of my change to the changelog in `HISTORY.md`
- [x] This PR is ready to go

### CODE REVIEW

File: ribs/emitters/_evolution_strategy_emitter.py [hunk 1]
```diff
@@ -193,6 +194,33 @@ def tell(self,
             metadata_batch (numpy.ndarray): 1D object array containing a
                 metadata object for each solution.
         """
+        # Shape checks
+        check_solution_batch_dim(solution_batch, "solution_batch",
```
Comment (btjanaka) [inline ribs/emitters/_evolution_strategy_emitter.py:198]: first check not needed as we allow returning variable number of solutions

File: tests/emitters/emitter_base_test.py [hunk 2]
```diff
@@ -71,6 +72,251 @@ def test_ask_emits_correct_num_sols_on_nonempty_archive(emitter_fixture):
...
+                                           x0=np.array([0]),
+                                           sigma0=1.0,
+                                           ranker="imp",
+                                           batch_size=batch_size)
+
+    return archive, emitter, batch_size
+
+
+@pytest.mark.parametrize("offset", [(0, 1), (1, 0), (1, 1)])
+def test_tell_wrong_shape_solution_batch(shape_test_fixture, offset):
```
Comment (btjanaka) [inline tests/emitters/emitter_base_test.py:109]: Parameterize with a parameter `wrong_array` that indicates which array to give a wrong shape to.

**Review by btjanaka — APPROVED** [review]

File: ribs/_utils.py [hunk 3]
```diff
@@ -60,3 +80,87 @@ def check_solution_batch_dim(array,
                          f"({batch_size}, ..), {array_name} should have shape "
                          f"({batch_size},{'' if is_1d else ' ..'}), but it has "
                          f"shape {array.shape}.{extra_msg}")
+
+
+_ADD_WARNING = (" Note that starting in pyribs 0.5.0, add() and tell() take"
```
Comment (btjanaka) [inline ribs/_utils.py:85]: should call this `_BATCH_WARNING`

File: ribs/_utils.py [hunk 4]
```diff
@@ -60,3 +80,87 @@ def check_solution_batch_dim(array,
...
+        metadata_batch = np.asarray(metadata_batch, dtype=object)
+        check_is_1d(metadata_batch, "metadata_batch", _ADD_WARNING)
+        check_solution_batch_dim(metadata_batch,
+                                 "metadata_batch",
+                                 batch_size,
+                                 is_1d=True,
+                                 extra_msg=_ADD_WARNING)
+
+
+def validate_add_single_args(archive, solution, objective, measures):
```
Comment (btjanaka) [inline ribs/_utils.py:159]: Is it only used in add_single, or can we call this validate_single_args?

Comment (btjanaka) [reply ribs/_utils.py:159]: And also rename validate_args to validate_batch_args

File: ribs/emitters/_evolution_strategy_emitter.py [hunk 5]
```diff
@@ -193,12 +193,21 @@ def tell(self,
             metadata_batch (numpy.ndarray): 1D object array containing a
                 metadata object for each solution.
         """
-        # Increase iteration counter.
-        self._itrs += 1
+        # Validate arguments.
+        validate_args(archive=self.archive,
```
Comment (btjanaka) [inline ribs/emitters/_evolution_strategy_emitter.py:197]: I think we should be calling asarray here too like in archivebase.add

File: ribs/emitters/_gradient_arborescence_emitter.py [hunk 6]
```diff
@@ -298,6 +298,16 @@ def tell_dqd(self,
             metadata_batch (numpy.ndarray): 1d object array containing a
                 metadata object for each solution.
         """
+        # Validate arguments.
+        validate_args(archive=self.archive,
```
Comment (btjanaka) [inline ribs/emitters/_gradient_arborescence_emitter.py:302]: same with asarray

File: tests/emitters/emitter_base_test.py [hunk 7]
```diff
@@ -71,6 +74,106 @@ def test_ask_emits_correct_num_sols_on_nonempty_archive(emitter_fixture):
...
+            ))
+        elif wrong_array == "objective_batch":
+            objective_batch = np.ones(batch_size + offset[0])
+        elif wrong_array == "measures_batch":
+            measures_batch = np.ones((
+                batch_size + offset[0],
+                archive.measure_dim + offset[1],
+            ))
+        elif wrong_array == "jacobian_batch":
+            print(offset)
```
Comment (btjanaka) [inline tests/emitters/emitter_base_test.py:144]: remove print

**Review by btjanaka — APPROVED** [review]
minor changes

File: ribs/_utils.py (hunk 4 above)
Comment (itsdawei) [reply ribs/_utils.py:159]: Its only used on add_single

**Pull request merged**